File: haskell/bob/Bob.hs
{-
Copyright 2014 Bas Bossink (bas.bossink@gmail.com).
-}

module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor a 
  | all isSpace a = "Fine. Be that way!" 
  | and [map toUpper a == a, any isLetter a] = "Woah, chill out!"
  | last a == '?' = "Sure."
  | otherwise = "Whatever."

File: scala/bob/src/main/scala/bob.scala
class Bob {
  def hey(poke: String): String = {
    val letters = poke.filter(_.isLetter)
    if (poke.filter(!_.isWhitespace).isEmpty) {
      "Fine. Be that way!"
    } else if (!letters.isEmpty && letters.forall(_.isUpper)) {
      "Woah, chill out!" 
    } else if (poke.last.equals('?') ) {
      "Sure." 
    } else {
      "Whatever."
    }
  }
}

